extensions-core/kinesis-indexing-service/src/main/java/org/apache/druid/indexing/kinesis/KinesisRecordSupplier.java

```diff
@@ -0,0 +1,782 @@
+                streamPartition.getStream(),
+                streamPartition.getPartitionId(),
+                kinesisRecord.getSequenceNumber(),
+                data
+            );
+
+
+            log.trace(
+                "Stream[%s] / partition[%s] / sequenceNum[%s] / bufferRemainingCapacity[%d]: %s",
+                currRecord.getStream(),
+                currRecord.getPartitionId(),
+                currRecord.getSequenceNumber(),
+                records.remainingCapacity(),
+                currRecord.getData().stream().map(StringUtils::fromUtf8).collect(Collectors.toList())
+            );
+
+            // If the buffer was full and we weren't able to add the message, grab a new stream iterator starting
+            // from this message and back off for a bit to let the buffer drain before retrying.
+            if (!records.offer(currRecord, recordBufferOfferTimeout, TimeUnit.MILLISECONDS)) {
+              log.warn(
+                  "OrderedPartitionableRecord buffer full, storing iterator and retrying in [%,dms]",
+                  recordBufferFullWait
+              );
+
+              shardIterator = kinesis.getShardIterator(
+                  currRecord.getStream(),
+                  currRecord.getPartitionId(),
+                  ShardIteratorType.AT_SEQUENCE_NUMBER.toString(),
+                  currRecord.getSequenceNumber()
+              ).getShardIterator();
+
+              rescheduleRunnable(recordBufferFullWait);
+              return;
+            }
+          }
+
+          shardIterator = recordsResult.getNextShardIterator(); // will be null if the shard has been closed
+
+          rescheduleRunnable(fetchDelayMillis);
+        }
+        catch (ProvisionedThroughputExceededException e) {
+          log.warn(
+              e,
+              "encounted ProvisionedThroughputExceededException while fetching records, this means "
+              + "that the request rate for the stream is too high, or the requested data is too large for "
+              + "the available throughput. Reduce the frequency or size of your requests."
+          );
+          long retryMs = Math.max(PROVISIONED_THROUGHPUT_EXCEEDED_BACKOFF_MS, fetchDelayMillis);
+          rescheduleRunnable(retryMs);
+        }
+        catch (InterruptedException e) {
+          // may happen if interrupted while BlockingQueue.offer() is waiting
+          log.warn(
+              e,
+              "Interrupted while waiting to add record to buffer, retrying in [%,dms]",
+              EXCEPTION_RETRY_DELAY_MS
+          );
+          if (currRecord != null) {
+            shardIterator = kinesis.getShardIterator(
+                currRecord.getStream(),
+                currRecord.getPartitionId(),
+                ShardIteratorType.AT_SEQUENCE_NUMBER.toString(),
+                currRecord.getSequenceNumber()
+            ).getShardIterator();
+            rescheduleRunnable(EXCEPTION_RETRY_DELAY_MS);
```

Comment:
Thanks for updating this. This looks good except that we should return immediately on InterruptedException.